In [11]:
#importamos primero seaborn y el df planetas 

import seaborn as sb

planets = sb.load_dataset("planets")


In [12]:
# analizamos el df

planets.shape

(1035, 6)

In [13]:
planets.head(10)

method  number  orbital_period   mass  distance  year
0  Radial Velocity       1         269.300   7.10     77.40  2006
1  Radial Velocity       1         874.774   2.21     56.95  2008
2  Radial Velocity       1         763.000   2.60     19.84  2011
3  Radial Velocity       1         326.030  19.40    110.62  2007
4  Radial Velocity       1         516.220  10.50    119.47  2009
5  Radial Velocity       1         185.840   4.80     76.39  2008
6  Radial Velocity       1        1773.400   4.64     18.15  2002
7  Radial Velocity       1         798.500    NaN     21.41  1996
8  Radial Velocity       1         993.300  10.30     73.10  2008
9  Radial Velocity       2         452.800   1.99     74.79  2010

In [14]:
planets.describe()  # como summary en R

number  orbital_period        mass     distance         year
count  1035.000000      992.000000  513.000000   808.000000  1035.000000
mean      1.785507     2002.917596    2.638161   264.069282  2009.070531
std       1.240976    26014.728304    3.818617   733.116493     3.972567
min       1.000000        0.090706    0.003600     1.350000  1989.000000
25%       1.000000        5.442540    0.229000    32.560000  2007.000000
50%       1.000000       39.979500    1.260000    55.250000  2010.000000
75%       2.000000      526.005000    3.040000   178.500000  2012.000000
max       7.000000   730000.000000   25.000000  8500.000000  2014.000000

Vemos que en count, tenemos cuantos datos hay, ecomo hay 1035 registros, habra NA ya que mass por ejemplo solo tiene 513 observaciones, para ello eliminamos dichos NA

In [15]:
planets.dropna().describe()

number  orbital_period        mass    distance         year
count  498.00000      498.000000  498.000000  498.000000   498.000000
mean     1.73494      835.778671    2.509320   52.068213  2007.377510
std      1.17572     1469.128259    3.636274   46.596041     4.167284
min      1.00000        1.328300    0.003600    1.350000  1989.000000
25%      1.00000       38.272250    0.212500   24.497500  2005.000000
50%      1.00000      357.000000    1.245000   39.940000  2009.000000
75%      2.00000      999.600000    2.867500   59.332500  2011.000000
max      6.00000    17337.500000   25.000000  354.000000  2014.000000

In [16]:
#vemos que drop na no borro datos
planets.describe()

number  orbital_period        mass     distance         year
count  1035.000000      992.000000  513.000000   808.000000  1035.000000
mean      1.785507     2002.917596    2.638161   264.069282  2009.070531
std       1.240976    26014.728304    3.818617   733.116493     3.972567
min       1.000000        0.090706    0.003600     1.350000  1989.000000
25%       1.000000        5.442540    0.229000    32.560000  2007.000000
50%       1.000000       39.979500    1.260000    55.250000  2010.000000
75%       2.000000      526.005000    3.040000   178.500000  2012.000000
max       7.000000   730000.000000   25.000000  8500.000000  2014.000000

In [17]:
# accedemos a var distancia y realizamos algunas operaciones
planets.distance



0        77.40
1        56.95
2        19.84
3       110.62
4       119.47
         ...  
1030    172.00
1031    148.00
1032    174.00
1033    293.00
1034    260.00
Name: distance, Length: 1035, dtype: float64

In [18]:
planets.distance.sum()


np.float64(213367.98000000004)

In [19]:
planets.distance.prod()

/home/miguel/anaconda3/lib/python3.13/site-packages/numpy/_core/_methods.py:57: RuntimeWarning: overflow encountered in reduce
  return umr_prod(a, axis, dtype, out, keepdims, initial, where)


np.float64(inf)

lo siguiente es la desvicion media absoluta
$$\frac {  \sum_{n=1}^n |x_i-\bar{x}|       }  {n}$$

In [20]:
# se calculaba con
#import pandas as ps
#planets.distance.mad()

# pero pandas ya no existe

In [21]:
# creemos la funcion
def mad(x):
    d = abs(x - x.mean()).mean()
    return d



In [22]:
mad(planets.distance)

np.float64(318.10088545240666)

In [23]:
planets.distance.var()

537459.7922212934

In [24]:
planets.distance.std()

733.1164929404422

 El .groupby en python es similar al aggregate o by, donde variable(s)~factor(es) solo que de la forma
 
df.groupby('factor(es)')['columna(s)'].FUN
 

In [25]:
planets.groupby('method')["orbital_period"].median()

method
Astrometry                         631.180000
Eclipse Timing Variations         4343.500000
Imaging                          27500.000000
Microlensing                      3300.000000
Orbital Brightness Modulation        0.342887
Pulsar Timing                       66.541900
Pulsation Timing Variations       1170.000000
Radial Velocity                    360.200000
Transit                              5.714932
Transit Timing Variations           57.011000
Name: orbital_period, dtype: float64

In [38]:
planets.groupby('method')["year"].describe()

count         mean       std     min      25%  \
method                                                                         
Astrometry                       2.0  2011.500000  2.121320  2010.0  2010.75   
Eclipse Timing Variations        9.0  2010.000000  1.414214  2008.0  2009.00   
Imaging                         38.0  2009.131579  2.781901  2004.0  2008.00   
Microlensing                    23.0  2009.782609  2.859697  2004.0  2008.00   
Orbital Brightness Modulation    3.0  2011.666667  1.154701  2011.0  2011.00   
Pulsar Timing                    5.0  1998.400000  8.384510  1992.0  1992.00   
Pulsation Timing Variations      1.0  2007.000000       NaN  2007.0  2007.00   
Radial Velocity                553.0  2007.518987  4.249052  1989.0  2005.00   
Transit                        397.0  2011.236776  2.077867  2002.0  2010.00   
Transit Timing Variations        4.0  2012.500000  1.290994  2011.0  2011.75   

                                  50%      75%     max  
method                                                  
Astrometry                     2011.5  2012.25  2013.0  
Eclipse Timing Variations      2010.0  2011.00  2012.0  
Imaging                        2009.0  2011.00  2013.0  
Microlensing                   2010.0  2012.00  2013.0  
Orbital Brightness Modulation  2011.0  2012.00  2013.0  
Pulsar Timing                  1994.0  2003.00  2011.0  
Pulsation Timing Variations    2007.0  2007.00  2007.0  
Radial Velocity                2009.0  2011.00  2014.0  
Transit                        2012.0  2013.00  2014.0  
Transit Timing Variations      2012.5  2013.25  2014.0

In [26]:
#otros ejemplos, en R c(year,orbital) ~ method 
planets.groupby('method')[["orbital_period",'year']].median()


orbital_period    year
method                                               
Astrometry                         631.180000  2011.5
Eclipse Timing Variations         4343.500000  2010.0
Imaging                          27500.000000  2009.0
Microlensing                      3300.000000  2010.0
Orbital Brightness Modulation        0.342887  2011.0
Pulsar Timing                       66.541900  1994.0
Pulsation Timing Variations       1170.000000  2007.0
Radial Velocity                    360.200000  2009.0
Transit                              5.714932  2012.0
Transit Timing Variations           57.011000  2012.5

In [34]:
planets.groupby(['year','method'])[["orbital_period"]].median()

orbital_period
year method                                   
1989 Radial Velocity                 83.888000
1992 Pulsar Timing                   45.901950
1994 Pulsar Timing                   98.211400
1995 Radial Velocity                  4.230785
1996 Radial Velocity                 65.669700
...                                        ...
2013 Transit                         18.164060
     Transit Timing Variations       22.339500
2014 Radial Velocity                820.300000
     Transit                          8.385385
     Transit Timing Variations             NaN

[69 rows x 1 columns]

In [27]:
#otros ejemplos
planets.groupby(['year','method'])[["orbital_period"]].median().unstack()


orbital_period                                                    \
method     Astrometry Eclipse Timing Variations    Imaging Microlensing   
year                                                                      
1989              NaN                       NaN        NaN          NaN   
1992              NaN                       NaN        NaN          NaN   
1994              NaN                       NaN        NaN          NaN   
1995              NaN                       NaN        NaN          NaN   
1996              NaN                       NaN        NaN          NaN   
1997              NaN                       NaN        NaN          NaN   
1998              NaN                       NaN        NaN          NaN   
1999              NaN                       NaN        NaN          NaN   
2000              NaN                       NaN        NaN          NaN   
2001              NaN                       NaN        NaN          NaN   
2002              NaN                       NaN        NaN          NaN   
2003              NaN                       NaN        NaN          NaN   
2004              NaN                       NaN        NaN          NaN   
2005              NaN                       NaN        NaN       3550.0   
2006              NaN                       NaN  730000.00       3300.0   
2007              NaN                       NaN        NaN          NaN   
2008              NaN                  4544.000   69000.00       3462.5   
2009              NaN                 10220.000    7336.50          NaN   
2010          1016.00                  4228.525   18000.00       1970.0   
2011              NaN                  2900.000   25018.75       2780.0   
2012              NaN                  4343.500        NaN          NaN   
2013           246.36                       NaN        NaN          NaN   
2014              NaN                       NaN        NaN          NaN   

                                                    \
method Orbital Brightness Modulation Pulsar Timing   
year                                                 
1989                             NaN           NaN   
1992                             NaN     45.901950   
1994                             NaN     98.211400   
1995                             NaN           NaN   
1996                             NaN           NaN   
1997                             NaN           NaN   
1998                             NaN           NaN   
1999                             NaN           NaN   
2000                             NaN           NaN   
2001                             NaN           NaN   
2002                             NaN           NaN   
2003                             NaN  36525.000000   
2004                             NaN           NaN   
2005                             NaN           NaN   
2006                             NaN           NaN   
2007                             NaN           NaN   
2008                             NaN           NaN   
2009                             NaN           NaN   
2010                             NaN           NaN   
2011                        0.291496      0.090706   
2012                             NaN           NaN   
2013                        1.544929           NaN   
2014                             NaN           NaN   

                                                               \
method Pulsation Timing Variations Radial Velocity    Transit   
year                                                            
1989                           NaN       83.888000        NaN   
1992                           NaN             NaN        NaN   
1994                           NaN             NaN        NaN   
1995                           NaN        4.230785        NaN   
1996                           NaN       65.669700        NaN   
1997                           NaN       39.845000        NaN   
1998                           NaN       58.112890        NaN   
1

In [37]:
# notemos que extraemos el metodo y el grupo del grupby, luego en prin el {0:30s} es para dar un espacio de 30 y alinearlos
for (method, group) in planets.groupby('method'):
    print("{0:30s} shape ={1}".format(method,group.shape))

Astrometry                     shape =(2, 6)
Eclipse Timing Variations      shape =(9, 6)
Imaging                        shape =(38, 6)
Microlensing                   shape =(23, 6)
Orbital Brightness Modulation  shape =(3, 6)
Pulsar Timing                  shape =(5, 6)
Pulsation Timing Variations    shape =(1, 6)
Radial Velocity                shape =(553, 6)
Transit                        shape =(397, 6)
Transit Timing Variations      shape =(4, 6)


# creando factores
supongamos que queremos abrupar por decada, para ello primero hay que crear dicha variable, tomando año, lo asignaremos a una decada




In [28]:
decada = planets.year//10 *10
decada

0       2000
1       2000
2       2010
3       2000
4       2000
        ... 
1030    2000
1031    2000
1032    2000
1033    2000
1034    2000
Name: year, Length: 1035, dtype: int64

In [42]:
planets.groupby(['method' , decada])[["number"]].sum()

number
method                        decada        
Astrometry                    2010s        2
Eclipse Timing Variations     2000s        5
                              2010s       10
Imaging                       2000s       29
                              2010s       21
Microlensing                  2000s       12
                              2010s       15
Orbital Brightness Modulation 2010s        5
Pulsar Timing                 1990s        9
                              2000s        1
                              2010s        1
Pulsation Timing Variations   2000s        1
Radial Velocity               1980s        1
                              1990s       52
                              2000s      475
                              2010s      424
Transit                       2000s       64
                              2010s      712
Transit Timing Variations     2010s        9

In [43]:
planets.groupby(['method' , decada])[["number"]].sum().unstack()

number                    
decada                         1980s 1990s  2000s  2010s
method                                                  
Astrometry                       NaN   NaN    NaN    2.0
Eclipse Timing Variations        NaN   NaN    5.0   10.0
Imaging                          NaN   NaN   29.0   21.0
Microlensing                     NaN   NaN   12.0   15.0
Orbital Brightness Modulation    NaN   NaN    NaN    5.0
Pulsar Timing                    NaN   9.0    1.0    1.0
Pulsation Timing Variations      NaN   NaN    1.0    NaN
Radial Velocity                  1.0  52.0  475.0  424.0
Transit                          NaN   NaN   64.0  712.0
Transit Timing Variations        NaN   NaN    NaN    9.0

In [44]:
planets.groupby(['method' , decada])[["number"]].sum().unstack().fillna(0)

number                    
decada                         1980s 1990s  2000s  2010s
method                                                  
Astrometry                       0.0   0.0    0.0    2.0
Eclipse Timing Variations        0.0   0.0    5.0   10.0
Imaging                          0.0   0.0   29.0   21.0
Microlensing                     0.0   0.0   12.0   15.0
Orbital Brightness Modulation    0.0   0.0    0.0    5.0
Pulsar Timing                    0.0   9.0    1.0    1.0
Pulsation Timing Variations      0.0   0.0    1.0    0.0
Radial Velocity                  1.0  52.0  475.0  424.0
Transit                          0.0   0.0   64.0  712.0
Transit Timing Variations        0.0   0.0    0.0    9.0

In [47]:
# termina clase
# incluso podriamos hacerlo como coloquialmente se hace con "80s"
decada = planets.year//10 *10
decada = decada.astype(str) +'s'
decada

0       2000s
1       2000s
2       2010s
3       2000s
4       2000s
        ...  
1030    2000s
1031    2000s
1032    2000s
1033    2000s
1034    2000s
Name: year, Length: 1035, dtype: object

In [48]:
planets.groupby([decada,'method'])[["orbital_period"]].sum().unstack()

orbital_period                                                     \
method     Astrometry Eclipse Timing Variations     Imaging Microlensing   
year                                                                       
1980s             NaN                       NaN         NaN          NaN   
1990s             NaN                       NaN         NaN          NaN   
2000s             NaN                   19308.0  1350935.35      17325.0   
2010s         1262.36                   23456.8    68037.50       4750.0   

                                                    \
method Orbital Brightness Modulation Pulsar Timing   
year                                                 
1980s                            NaN           NaN   
1990s                            NaN    190.015300   
2000s                            NaN  36525.000000   
2010s                        2.12792      0.090706   

                                                                 \
method Pulsation Timing Variations Radial Velocity      Transit   
year                                                              
1980s                          NaN       83.888000          NaN   
1990s                          NaN    10915.612813          NaN   
2000s                       1170.0   263352.646027   289.710182   
2010s                          NaN   180962.991210  8087.812669   

                                  
method Transit Timing Variations  
year                              
1980s                        NaN  
1990s                        NaN  
2000s                        NaN  
2010s                   239.3505

In [49]:
# notemos que la va decada no tiene mas que el nombre heredado del df year, por tanto podemos cambiarlo con
decada.name = "decada"
planets.groupby([decada,'method'])[["orbital_period"]].sum().unstack()

orbital_period                                                     \
method     Astrometry Eclipse Timing Variations     Imaging Microlensing   
decada                                                                     
1980s             NaN                       NaN         NaN          NaN   
1990s             NaN                       NaN         NaN          NaN   
2000s             NaN                   19308.0  1350935.35      17325.0   
2010s         1262.36                   23456.8    68037.50       4750.0   

                                                    \
method Orbital Brightness Modulation Pulsar Timing   
decada                                               
1980s                            NaN           NaN   
1990s                            NaN    190.015300   
2000s                            NaN  36525.000000   
2010s                        2.12792      0.090706   

                                                                 \
method Pulsation Timing Variations Radial Velocity      Transit   
decada                                                            
1980s                          NaN       83.888000          NaN   
1990s                          NaN    10915.612813          NaN   
2000s                       1170.0   263352.646027   289.710182   
2010s                          NaN   180962.991210  8087.812669   

                                  
method Transit Timing Variations  
decada                            
1980s                        NaN  
1990s                        NaN  
2000s                        NaN  
2010s                   239.3505

In [50]:
# quitemos los na
planets.groupby([decada,'method'])[["number"]].sum().unstack().fillna(0)

number                                                 \
method Astrometry Eclipse Timing Variations Imaging Microlensing   
decada                                                             
1980s         0.0                       0.0     0.0          0.0   
1990s         0.0                       0.0     0.0          0.0   
2000s         0.0                       5.0    29.0         12.0   
2010s         2.0                      10.0    21.0         15.0   

                                                    \
method Orbital Brightness Modulation Pulsar Timing   
decada                                               
1980s                            0.0           0.0   
1990s                            0.0           9.0   
2000s                            0.0           1.0   
2010s                            5.0           1.0   

                                                            \
method Pulsation Timing Variations Radial Velocity Transit   
decada                                                       
1980s                          0.0             1.0     0.0   
1990s                          0.0            52.0     0.0   
2000s                          1.0           475.0    64.0   
2010s                          0.0           424.0   712.0   

                                  
method Transit Timing Variations  
decada                            
1980s                        0.0  
1990s                        0.0  
2000s                        0.0  
2010s                        9.0